ERCS/erc-7546.md

```diff
@@ -0,0 +1,187 @@
+---
+eip: 7546
+title: Upgradeable Clone
+description: An extendible and cheaper proxy pattern.
+author: Shogo Ochiai (@shogochiai) <shogo.ochiai@pm.me>, Kai Hiroi (@KaiHiroi) <kai.hiroi@pm.me>
+discussions-to: https://ethereum-magicians.org/t/eip-7546-upgradeable-clone/16256
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-10-25
+requires: 165, 1967, 7201
+---
+
+## Abstract
+This ERC proposes an upgradeable proxy standard that's compatible with the factory (clone) pattern.
+
+This standard is comprised of the following:
+1. **Dictionary Contract** that manages implementation contract addresses paired with function selectors.
+2. **Proxy Contract** that delegatecalls to the implementation address registered in the Dictionary.
+3. **Interface** that represents the behavior of the Proxy as a collection of implementations.
+4. **Storage Layout** that ensures no conflicts when shared among multiple implementations.
+
+
+## Motivation
+This standard caters to:
+
+1. Users seeking the factory pattern for cloning contracts with common, upgradeable implementations.
+2. Scenarios requiring simultaneous upgrades of cloned contracts.
+3. Desires for flexible, per-function upgradeability.
+
+
+## Specification
+> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### 1. Dictionary Contract
+This contract manages a mapping of function selectors to corresponding implementation addresses to cater to calls from the Proxy.
+
+#### 1.1 Storage & Events
+The Dictionary MUST maintain a mapping of function selectors to implementation addresses.
+
+Changes to this mapping SHOULD be communicated through an event (or log).
+```solidity
+event ImplementationUpgraded(bytes4 indexed functionSelector, address indexed implementation);
+```
+
+#### 1.2 Functions
+The Dictionary SHOULD provide functions to retrieve and set implementations.
+
+##### 1.2.1 `getImplementation`
+```solidity
+function getImplementation(bytes4 functionSelector) external view returns (address);
+```
+
+##### 1.2.2 `setImplementation`
+```solidity
+function setImplementation(bytes4 functionSelector, address implementation) external;
+```
+The implementation contract MUST implement the function selector to be registered as a key in the mapping (section 3.1). Therefore, it is RECOMMENDED to verify this during registration.
```

Comment:
I don't understand this requirement. What does "implement the function selector to be registered" mean? What needs to be verified?

Reply:
Thank you for your feedback. We have added the content and moved it to the Security Considerations section.